python/dgl/nn/pytorch/conv.py

```diff
@@ -148,3 +148,219 @@ def extra_repr(self):
         if '_activation' in self.__dict__:
             summary += ', activation={_activation}'
         return summary.format(**self.__dict__)
+
+class _BaseRelGraphConv(nn.Module):
+    """Base Relational graph convolution layer.
+
+    Parameters
+    ----------
+    in_feat : int
+        The input feature size
+    out_feat : int
+        The output feature size
+    bias : bool, optional
+        True if bias is added.
+    activation : callable, optional
+        Activation function.
+    self_loop : bool, optional
+        True to include self loop message.
+    dropout : float, optional
+        Dropout rate.
+    """
+    def __init__(self,
+                 in_feat,
+                 out_feat,
+                 bias,
+                 activation,
+                 self_loop,
+                 dropout):
+        super(_BaseRelGraphConv, self).__init__()
+        self.bias = bias
+        self.activation = activation
+        self.self_loop = self_loop
+
+        if self.bias:
+            self.h_bias = nn.Parameter(th.Tensor(out_feat))
+            nn.init.zeros_(self.h_bias)
+
+        # weight for self loop
+        if self.self_loop:
+            self.loop_weight = nn.Parameter(th.Tensor(in_feat, out_feat))
+            nn.init.xavier_uniform_(self.loop_weight,
+                                    gain=nn.init.calculate_gain('relu'))
+
+        self.dropout = nn.Dropout(dropout)
+
+    def message_func(self, edges):
+        """Message function that should be implemented by the subclass."""
+        raise NotImplemented()
+
+    def forward(self, g, h, r, norm=None):
+        """Forward computation
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph.
+        h : torch.Tensor
+            Input node features. Shape: (|V|, D)
+        r : torch.Tensor
+            Edge type tensor. Shape: (|E|,)
+        norm : torch.Tensor
+            Optional edge normalizer tensor. Shape: (|E|, 1)
+
+        Returns
+        -------
+        torch.Tensor
+            New node features.
+        """
+        g = g.local_var()
+        g.ndata['h'] = h
+        g.edata['type'] = r
+        if norm is not None:
+            g.edata['norm'] = norm
+        if self.self_loop:
+            loop_message = th.mm(g.ndata['h'], self.loop_weight)
+            loop_message = self.dropout(loop_message)
+
+        # message passing
+        g.update_all(self.message_func, fn.sum(msg='msg', out='h'))
+
+        # apply bias and activation
+        node_repr = g.ndata['h']
+        if self.bias:
+            node_repr = node_repr + self.h_bias
+        if self.self_loop:
+            node_repr = node_repr + loop_message
+        if self.activation:
+            node_repr = self.activation(node_repr)
+
+        return node_repr
+
+class RelGraphConvBasis(_BaseRelGraphConv):
+    """Relational Graph Convolution with basis regularization.
```

Comment:
"""Apply relational graph convolution with basis regularization graph convolution over an input signal. Relational graph convolution is introduced in `RGCN <https://arxiv.org/abs/1703.06103>`__
    and can be described as below: